File: books-service/src/controllers/healthController.js
// ============================================================
// src/controllers/healthController.js
// Three health endpoints:
//   GET /health        — service alive (process up)
//   GET /health/live   — liveness probe (process not stuck)
//   GET /health/ready  — readiness probe (PostgreSQL reachable)
// ============================================================
'use strict';

const db     = require('../config/database');
const config = require('../config/config');
const logger = require('../utils/logger');

const META = {
  service: config.app.name,
  version: config.app.version,
};

// ── GET /health ───────────────────────────────────────────────
async function health(req, res) {
  return res.status(200).json({
    status:    'UP',
    message:   'Books service is running',
    ...META,
    timestamp: new Date().toISOString(),
  });
}

// ── GET /health/live ─────────────────────────────────────────
// Liveness — is the Node process alive?
// Returns 200 if process is alive, 503 if it should be restarted.
async function liveness(req, res) {
  return res.status(200).json({
    status:    'UP',
    probe:     'liveness',
    message:   'Application process is alive',
    ...META,
    timestamp: new Date().toISOString(),
  });
}

// ── GET /health/ready ─────────────────────────────────────────
// Readiness — is the service ready to accept traffic?
// Checks PostgreSQL connectivity. Returns 503 if DB is down.
async function readiness(req, res) {
  const checks = {};
  let allHealthy = true;

  try {
    await db.ping();
    checks.postgresql = 'UP';
  } catch (err) {
    logger.error('Health check: PostgreSQL DOWN', { error: err.message });
    checks.postgresql = `DOWN - ${err.message}`;
    allHealthy = false;
  }

  const statusCode = allHealthy ? 200 : 503;
  return res.status(statusCode).json({
    status:    allHealthy ? 'UP' : 'DOWN',
    probe:     'readiness',
    message:   allHealthy
      ? 'Application is ready to serve traffic'
      : 'Application is NOT ready — dependencies unavailable',
    checks,
    ...META,
    timestamp: new Date().toISOString(),
  });
}

module.exports = { health, liveness, readiness };

File: payment-service/src/main/java/com/emart/payment/controller/HealthController.java
package com.emart.payment.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Three health endpoints — NO authentication required.
 *
 * GET /health        → process alive check
 * GET /health/live   → liveness probe (K8s: restart if fails)
 * GET /health/ready  → readiness probe (K8s: remove from LB if fails)
 */
@RestController
@RequiredArgsConstructor
@Slf4j
public class HealthController {

    private final JdbcTemplate jdbcTemplate;

    @Value("${spring.application.name:emart-payment-service}")
    private String serviceName;

    @Value("${app.version:1.0.0}")
    private String version;

    // ── GET /health ───────────────────────────────────────────
    @GetMapping("/health")
    public ResponseEntity<Map<String, Object>> health() {
        return ResponseEntity.ok(Map.of(
            "status",    "UP",
            "service",   serviceName,
            "version",   version,
            "message",   "Payment service is running",
            "timestamp", Instant.now().toString()
        ));
    }

    // ── GET /health/live ──────────────────────────────────────
    @GetMapping("/health/live")
    public ResponseEntity<Map<String, Object>> liveness() {
        return ResponseEntity.ok(Map.of(
            "status",    "UP",
            "probe",     "liveness",
            "service",   serviceName,
            "message",   "Application process is alive",
            "timestamp", Instant.now().toString()
        ));
    }

    // ── GET /health/ready ─────────────────────────────────────
    @GetMapping("/health/ready")
    public ResponseEntity<Map<String, Object>> readiness() {
        Map<String, Object> checks  = new LinkedHashMap<>();
        boolean             healthy = true;

        // PostgreSQL connectivity check
        try {
            jdbcTemplate.queryForObject("SELECT 1", Integer.class);
            checks.put("postgresql", "UP");
        } catch (Exception ex) {
            log.error("Readiness: PostgreSQL check failed: {}", ex.getMessage());
            checks.put("postgresql", "DOWN");
            healthy = false;
        }

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("status",    healthy ? "UP" : "DOWN");
        body.put("probe",     "readiness");
        body.put("service",   serviceName);
        body.put("message",   healthy
            ? "Application is ready to serve traffic"
            : "Application is NOT ready — dependencies unavailable");
        body.put("checks",    checks);
        body.put("timestamp", Instant.now().toString());

        return ResponseEntity.status(healthy ? 200 : 503).body(body);
    }
}
